framework/base/Module.php

```diff
@@ -512,29 +534,69 @@ public function setModules($modules)
      * This method parses the specified route and creates the corresponding child module(s), controller and action
      * instances. It then calls [[Controller::runAction()]] to run the action with the given parameters.
      * If the route is empty, the method will use [[defaultRoute]].
+     * @param Request $request the request instance.
      * @param string $route the route that specifies the action.
      * @param array $params the parameters to be passed to the action
      * @return mixed the result of the action.
      * @throws InvalidRouteException if the requested route cannot be resolved into an action successfully.
      */
-    public function runAction($route, $params = [])
+    public function runAction($request, $route, $params = [])
```

Comment:
Could be type-hinted with `Request`.

Reply:
Unlikely.
PSR middleware does not provide any garantee the original request instance to be preserved. It passes `ServerRequestInterface` instance over the tiers, but can modify it in process or even recreate at will.

```
public function process(ServerRequestInterface $request, RequestHandlerInterface $handler)
{
    $newRequst = new ExternalRequest();
    return $handler->handle($newRequst);
}
```

Thus for the web application the correct typehinting will be `ServerRequestInterface`, while for the console application it will be `yii\console\Request` (or `yii\base\Request`).

`yii\base\Module` is suitable both for web and console application, thus such bounds can not be applied without introduction of the separated `yii\web\Module` and `yii\console\Module`. Such change will cause a lot of trouble. In particular: `yii2-gii` will have to implement 2 separated modules for the web and console interfaces, which then will require separated configuration.

Reply:
> It passes ServerRequestInterface instance over the tiers, but can modify it in process or even recreate at will.

Yes but what's returned is still `ServerRequestInterface`, isn't it?

Reply:
Ah. I've got what you mean. Console. Alright then.